modules/apps/product-navigation/product-navigation-control-menu-web/src/main/resources/META-INF/resources/js/useDragAndDrop.js

```diff
@@ -0,0 +1,312 @@
+		},
+	});
+
+	useEffect(() => {
+		previewRef(getEmptyImage(), {captureDraggingState: true});
+	}, [previewRef]);
+
+	return {
+		handlerRef,
+		isDraggingSource,
+		sourceRef,
+	};
+}
+
+export function useDragSymbol({data, icon, label, portletId, type}) {
+	const sourceItem = useMemo(
+		() => ({
+			data,
+			icon,
+			isSymbol: true,
+			itemId: portletId,
+			name: label,
+			type,
+		}),
+		[data, icon, label, portletId, type]
+	);
+
+	const {handlerRef, isDraggingSource, sourceRef} = useDragItem(sourceItem);
+
+	const symbolRef = (element) => {
+		sourceRef.current = element;
+		handlerRef(element);
+	};
+
+	return {
+		isDraggingSource,
+		sourceRef: symbolRef,
+	};
+}
+
+export function useDropClear(targetItem) {
+	const {dropTargetColumn, setDropTargetColumn} = useContext(
+		DragAndDropContext
+	);
+
+	const [, setDropClearRef] = useDrop({
+		accept: Object.values(LAYOUT_DATA_ITEM_TYPES),
+		hover(item, monitor) {
+			if (!monitor.isOver({shallow: true})) {
+				return;
+			}
+
+			if (dropTargetColumn) {
+				dropTargetColumn.classList.remove(DROP_OVER_CLASS);
+			}
+
+			setDropTargetColumn(null);
+		},
+	});
+
+	setDropClearRef(targetItem);
+}
+
+export function useDropTarget(targetItem) {
+	const {
+		dropTargetColumn,
+		dropTargetItem,
+		setDragIndicatorPosition,
+		setDropTargetColumn,
+		setDropTargetItem,
+	} = useContext(DragAndDropContext);
+
+	const [windowScrollPosition, setWindowScrollPosition] = useState(0);
+	const [targetPosition, setTargetPosition] = useState(null);
+
+	const setWidgets = useSetWidgetsContext();
+	const widgets = useWidgetsContext();
+	const plid = usePlidContext();
+
+	useEffect(() => {
+		const handleWindowScroll = () => {
+			setWindowScrollPosition(window.scrollY);
+		};
+
+		window.addEventListener('scroll', handleWindowScroll);
+
+		return () => {
+			window.removeEventListener('scroll', handleWindowScroll);
+		};
+	}, []);
+
+	const [, setDropTargetRef] = useDrop({
+		accept: Object.values(LAYOUT_DATA_ITEM_TYPES),
+		drop(item, monitor) {
+			setDropTargetColumn(null);
+
+			if (!monitor.isOver({shallow: true})) {
+				return;
+			}
+
+			if (!item.disabled) {
+				dropTargetColumn.classList.remove(DROP_OVER_CLASS);
+
+				addPortlet({item, plid, targetItem, targetPosition});
+
+				if (!item.data.instanceable) {
+					const updatedWidgets = updateUsedWidget({
+						item,
+						widgets,
+					});
+					setWidgets(updatedWidgets);
+				}
+			}
+		},
+		hover(item, monitor) {
+			if (!monitor.isOver({shallow: true})) {
+				return;
+			}
+
+			const itemIsDropzone = targetItem.classList.contains(
+				'portlet-dropzone'
+			);
+
+			const parentTargetItem = itemIsDropzone
+				? targetItem.parentElement
+				: targetItem.parentElement.parentElement;
+
+			if (dropTargetColumn !== parentTargetItem) {
+				if (dropTargetColumn) {
+					dropTargetColumn.classList.remove(DROP_OVER_CLASS);
+				}
+				setDropTargetColumn(parentTargetItem);
+				parentTargetItem.classList.add(DROP_OVER_CLASS);
```

Comment:
Wrapping this in `if (itemIsDropzone) {}` we could avoid the double feedback when hovering 😃